<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgres://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*)
FROM stores


"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT name
FROM stores
ORDER BY name
LIMIT 4
"""

pd.read_sql_query(sql, engine)

name
0  218 Fuel Express & Chubby's Liquor
1          3047 Cub Foods / Iowa City
2               3060 Cub Foods / Ames
3         3061 Cub Foods / Sioux City

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price, bottle_qty, total
FROM sales AS s
WHERE s.category_name = 'IMPORTED VODKA'
ORDER BY vendor 
LIMIT 3
"""

pd.read_sql_query(sql, engine)

btl_price  bottle_qty   total
0     $8.99          12  107.88
1    $28.50           6  171.00
2    $28.50           1   28.50

**Prompt 3 Answer**: 


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(bottle_qty) tot_bottles, SUM(total) tot_$
FROM sales
WHERE date between '2015-02-01' and '2015-02-28'
LIMIT 20


"""
pd.read_sql_query(sql, engine)

tot_bottles        tot_$
0      1591962  21295350.55

**Prompt 4 Answer**: 

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*) transactions, SUM(bottle_qty) bottles, SUM(total) total_$, vendor
FROM sales AS s
WHERE s.category_name = 'IMPORTED VODKA'
GROUP BY s.vendor
ORDER BY COUNT(*) DESC
LIMIT 1

"""

pd.read_sql_query(sql, engine)

transactions  bottles     total_$                            vendor
0         42338   403450  7880365.88  Pernod Ricard USA/Austin Nichols

**Prompt 5 Answer**: 

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT (btl_price - state_btl_cost) markup
FROM sales
ORDER BY markup DESC
LIMIT 10

"""

pd.read_sql_query(sql, engine)

markup
0  $2,900.00
1  $2,900.00
2    $799.60
3    $799.60
4    $799.60
5    $799.60
6    $799.60
7    $799.60
8    $799.60
9    $799.60

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT AVG( (btl_price - state_btl_cost) * 100 / state_btl_cost) markup_percent, store
FROM sales
GROUP BY store
ORDER BY markup_percent DESC
LIMIT 5


"""

pd.read_sql_query(sql, engine)

markup_percent  store
0       52.547124   4013
1       52.149880   4024
2       51.890342   4266
3       51.509378   3833
4       51.459084   3660

**Prompt 7 Answer**: 

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [9]:
# Use a SQL query to find the answer to the above prompt.


sql = """

SELECT  AVG( (sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost )*100 AS markup_percent,
        MIN((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost )*100 AS min_markup,
        MAX((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost )*100 AS max_markup,
        sa.store, st.name, st.store_address
FROM sales AS sa
LEFT JOIN stores AS st ON st.store = sa.store
GROUP BY st.name, st.store_address, sa.store
ORDER BY markup_percent DESC
LIMIT 5

"""


pd.read_sql_query(sql, engine)

markup_percent  min_markup  max_markup  store                        name  \
0       52.547124   49.857143  670.876289   4013      Wal-Mart 0841 / Tipton   
1       52.149880   49.846154  670.876289   4024  Wal-Mart 1546 / Iowa Falls   
2       51.890342   49.846154  670.876289   4266  Wal-Mart 1683 / Shenandoah   
3       51.509378   49.846154  670.876289   3833    Wal-Mart 3394 / Atlantic   
4       51.459084   49.846154  670.876289   3660   Wal-Mart 2935 / Knoxville   

                                       store_address  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...

**Prompt 8 Answer**:

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [10]:
# Use a SQL query to find the answer to the above prompt.

# TOP 9 COUNTIES with HIGHEST # of bottles sold/person
# AVG (bottle size) rounded to the tenth. 
# ORDER BY AVG (bottle size) descending


sql = """

SELECT 
    
    SUM(s.bottle_qty)/c.population AS btl_person, 
    ROUND(AVG(p.bottle_size),1) AS bottle_size ,
    s.county
FROM sales AS s
LEFT JOIN products AS p 
    ON s.item = p.item_no
RIGHT JOIN counties AS c
    ON s.county = c.county

GROUP BY s.county, c.population
ORDER BY btl_person DESC 
LIMIT 9


"""

pd.read_sql_query(sql, engine)

btl_person  bottle_size         county
0          22        974.1      Dickinson
1          15        862.9     Black Hawk
2          15        892.6           Polk
3          14        986.4    Cerro Gordo
4          13        895.4        Johnson
5          13        871.6          Scott
6          12        886.3           Linn
7          11        893.7  Pottawattamie
8          11       1041.0        Kossuth

**Prompt 9 Answer**: KOSSUTH

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT  *, (btl_price * bottle_qty)
FROM sales
WHERE (btl_price * bottle_qty) != MONEY(total)
LIMIT 100


"""

pd.read_sql_query(sql, engine)

Empty DataFrame
Columns: [date, convenience_store, store, county_number, county, category, category_name, vendor_no, vendor, item, description, pack, liter_size, state_btl_cost, btl_price, bottle_qty, total, ?column?]
Index: []

**Prompt 10 Answer**: It appears there are no transactions. ZERO
